domain_coordinator/domain_coordinator/__init__.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2019 Apex.AI, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import socket
+
+# To coordinate ROS_DOMAIN_IDs between multiple test process instances, we
+# open a high numbered port "PORT_BASE + ROS_DOMAIN_ID."
+# If we manage to open the port, then we can use that ROS_DOMAIN_ID for the duration of the
+# test run
+_PORT_BASE = 22119  # I picked this randomly as a high port that probably won't be in use
+
+
+class _sockwrapper():
+    """Wraps sockets to keep them open, but appear like a number from 1 to 100."""
+
+    def __init__(self, socket):
+        self.__socket = socket
+
+    def __str__(self):
+        return str(self.__socket.getsockname()[1] - _PORT_BASE)
+
+
+class _default_selector:
+
+    def __init__(self):
+        # When we need to coordinate 10 or 20 domains, it's about 10x faster
+        # to start with a random seed value here.
+        # It's also the difference between 1ms and 100us so it's totally insignificant.
+        # Leaving this here in case it's ever useful in the future to speed up domain selection:
+        # self._value = random.randint(1, 100)
+
+        # Slower, but deterministic:
+        # Always start at '1' so if there's weirdness where domains are colliding when
+        # they shouldn't, it's easier to debug.
+        self._value = 1
+
+    def __call__(self):
+        retval = ((self._value - 1) % 100) + 1
+        self._value += 1
+        return retval
+
+
+def get_coordinated_domain_id(*, selector=None):
+    """
+    Get a ROS_DOMAIN_ID from 1 to 100 that will not conflict with other ROS_DOMAIN_IDs.
+
+    Processes can use get_coordinated_domain_id to generate ROS_DOMAIN_IDs that allow them to
+    use ROS2 without unexpected cross-talk between processes.
+    This is similar to the ROS1 rostest behavior of putting the ROS master on a unique port.
+
+    Users of get_coordintaed_gomain_id must keep the returned object alive.  If the returned
```

Comment:
This is conceptionally really broken. In Python you can't rely on when an instance which isn't referenced anymore is being garbage collected. As a result the tests even warn about the sockets never being closed: https://ci.ros2.org/view/All/job/test_ci_windows/196/pytest-warnings/

See #9 for the follow up ticket.

Reply:
@dirk-thomas The open port shouldn't outlive the `run_test_isolated.py` wrapper.  The port will be returned to the pool of valid ports when `run_test_isolated.py` is terminated

Reply:
Obviously the resource doesn't outlive the process lifetime. But that doesn't solve the API problem which in its current form results in warnings about unclosed handles.

Reply:
Yeah - sorry, I edited at the same time you replied.  The API documentation does kind of write a check that the implementation can't cash right now.  I also agree that useless warnings are bad noise.